include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -887,6 +887,146 @@ class discard_iterator
     difference_type __my_position_;
 };
 
+template <typename _Iter>
+class direct_iterator
+{
+  public:
+    typedef typename ::std::iterator_traits<_Iter>::value_type value_type;
+    typedef typename ::std::iterator_traits<_Iter>::difference_type difference_type;
+    typedef typename ::std::iterator_traits<_Iter>::reference reference;
+
+    using iterator_category = std::random_access_iterator_tag;
+
+    using iterator = direct_iterator<_Iter>;
+    using pointer = iterator;
+
+    using is_passed_directly = ::std::true_type;
+
+    direct_iterator(_Iter iter) noexcept : __iter(iter) {}
+    direct_iterator() noexcept = default;
+    ~direct_iterator() noexcept = default;
+    direct_iterator(const direct_iterator&) noexcept = default;
+    direct_iterator&
+    operator=(const direct_iterator&) noexcept = default;
+
+    bool
+    operator==(const direct_iterator& other) const noexcept
+    {
+        return __iter == other.__iter;
+    }
+    bool
+    operator!=(const direct_iterator& other) const noexcept
+    {
+        return __iter != other.__iter;
+    }
+
+    iterator
+    operator+(difference_type offset) const noexcept
+    {
+        return iterator(__iter + offset);
+    }
+
+    iterator
+    operator-(difference_type offset) const noexcept
+    {
+        return iterator(__iter - offset);
+    }
+
+    difference_type
+    operator-(iterator other) const noexcept
+    {
+        return __iter - other.__iter;
+    }
+
+    bool
+    operator<(iterator other) const noexcept
+    {
+        return __iter < other.__iter;
+    }
+
+    bool
+    operator>(iterator other) const noexcept
+    {
+        return __iter > __iter;
+    }
+
+    bool
+    operator<=(iterator other) const noexcept
+    {
+        return __iter <= other.__iter;
+    }
+
+    bool
+    operator>=(iterator other) const noexcept
+    {
+        return __iter >= other.__iter;
+    }
+
+    iterator&
+    operator++() noexcept
+    {
+        ++__iter;
+        return *this;
+    }
+
+    iterator
+    operator++(int) noexcept
+    {
+        iterator other = *this;
+        ++(*this);
+        return other;
+    }
+
+    iterator&
+    operator--() noexcept
+    {
+        --__iter;
+        return *this;
+    }
+
+    iterator
+    operator--(int) noexcept
+    {
+        iterator other = *this;
+        --(*this);
+        return other;
+    }
+
+    iterator&
+    operator+=(difference_type offset) noexcept
+    {
+        __iter += offset;
+        return *this;
+    }
+
+    iterator&
+    operator-=(difference_type offset) noexcept
+    {
+        __iter -= offset;
+        return *this;
+    }
+
+    reference operator*() const noexcept { return *__iter; }
```

Comment:
To tell the truth, I really don't  understand an essence of that wrapper over `_Iter`.
That wrapper repeats the all standard RA iterator functionality, including dereferencing. If `_Iter` is not accessible on a device, `direct_iterator` also is not accessible on a device... So, what's an essence here?

Reply:
@MikeDvorskiy Sorry for being late getting back to you; this slipped past my inbox.

The idea here is that you have a range/iterator accessible on the device.  Let's say a `std::span<int>` to device memory.  Then, you create a view based on that range.  For example:

```cpp
template <typename T>
auto sum_times_two(std::span<T> x) {
  auto z = x
         | std::ranges::views::transform(
           [](auto &&elem) {
             return elem*2;
           });

  oneapi::dpl::execution::device_policy policy(q);

  return oneapi::dpl::experimental::reduce_async(
             policy, z.begin(), z.end(), T(0), std::plus())
      .get();
}
```

This code works, but has terrible performance.  The reason is that oneDPL does not know `transform_view<...>::iterator` is device accessible, so it copies all the elements one-by-one from the device to the host, then uses a buffer to copy it back to the device.  We use `direct_iterator` to force oneDPL to use the iterator directly, since we know that it can be used directly on the device.

```cpp
template <typename T>
auto sum_times_two(std::span<T> x) {
  auto z = x
         | std::ranges::views::transform(
           [](auto &&elem) {
             return elem*2;
           });

  oneapi::dpl::execution::device_policy policy(q);

  shp::__detail::direct_iterator d_first(z.begin());
  shp::__detail::direct_iterator d_last(z.end());

  return oneapi::dpl::experimental::reduce_async(
             policy, d_first, d_last, T(0), std::plus())
      .get();
}
```

This example is a bit simplified.  In the use case in distributed ranges, we have an actual `device_ptr` as the underlying iterator type, so we do know that the data lives on the device.  It might be worth thinking about how we could integrate distributed range's concepts of device vs. host memory with distributed ranges, but I think there will always be some cases where a user wants to explicitly "promote" a range to being directly accessible on the device.  Using a standard library view is a prime example of this, as we're unlikely to be able to hardwire locality information into a view without modifying the standard.  (Or providing our own implementation of all views.)